rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+
+Naming these functions to be clearly understood will be important so "Anti-Chaos-Monkey" can't be what we go with here. ;) The current proposal for the name for the "Write-Only Automatic" style functions is "Enforce" as the user is asking IaSQL to enforce these particular rows or tables to always match the database, and the "Read-Only Automatic" style becomes "Track" as the user is asking IaSQL to track the upstream cloud. The proposed new syntax for these is:
+
+```sql
+SELECT iasql_enforce('table_name'); -- Enforce an entire table onto the cloud.
+
+SELECT iasql_enforce('table_name', 5); -- Enforce a particular row from a table onto the cloud. We need to use the database ID here because enforcement *will* cause `@cloudId` churn if it needs to recreate the resource.
+
+SELECT iasql_enforce_all(); -- Essentially makes the DB write-only
+
+SELECT iasql_drop_enforce('table_name'); -- Drop the enforcement of the table. Should it also automatically drop specific row-level enforcement is something to decide. I'm leaning towards "yes" but it could be unexpected.
+
+SELECT iasql_drop_enforce('table_name', 5); -- Drop the enforcement of a particular row. This one, at least, is unambiguous.
+
+SELECT iasql_drop_enforce_all(); -- Restores normal behavior.
+
+SELECT iasql_track('table_name'); -- Makes the table effectively read-only.
+
+SELECT iasql_track('table_name', 5); -- Makes a particular row read-only.
+
+SELECT iasql_track_all(); -- Makes the DB read-only
+
+SELECT iasql_drop_track('table_name'); -- Drops tracking for the table. Similar issue to above: should this affect row-level tracking in the same table?
+
+SELECT iasql_drop_track('table_name', 5); -- Drops tracking for a particular row.
+
+SELECT iasql_drop_track_all(); -- Restores normal behavior
+```
+
+While entering the transaction mode introduces a synchronous point (such that `CALL iasql_commit();` blocks until the apply/sync operation is complete), there may be times where a user made changes in "normal" two-way mode but they now need to wait for resources to actually be created. A proposed helper function would trigger a blocking call until a user condition is met. One possible syntax for this is:
+
+```sql
+SELECT iasql_wait_for('table_name', 5, 'cloud_id IS NOT NULL'); -- Default version
+SELECT iasql_wait_for('table_name', 5, 'cloud_id IS NOT NULL', 30); -- Optional argument for maximum # of seconds to wait before erroring (defaulting to 5 minutes or something)
+```
```

Comment:
hmm, but isn't the `iasql_commit` call blocking? I'd imagine it like this: call commit, execute all the changes, enable the cron, and exit.

Reply:
ohh forget it, I understand now that the `isql_wait_for` is for the cases when no transaction like was executed, but normal 2way mode

Reply:
Yep :)

Reply:
Yes! We should also have a `iasql_wait_for_all()`? Probably from the get-go?

Reply:
What would `wait_for_all` do?

Reply:
I mean wait for all apply/sync transactions to complete